File: backend/src/phasor_workbench/serialization/paths.py
"""Translating a Pydantic error location into a place in the raw document.

Pydantic reports where a problem is as a `loc` tuple, and that tuple is **not** a
path into the document. For a discriminated union it names the branch it matched:
an illegal `controls` on a barrier reports

    ('operations', 1, 'barrier', 'controls')

where the document has no `barrier` key at all. Anything indexing the document
with a raw `loc` walks off the structure, and anything formatting one as a path
tells the client to look at a field it never sent.

Paths use the format in docs/API.md, matching `validation/paths.py`, so the
project has one vocabulary for locating a thing inside a circuit.
"""

from __future__ import annotations

from collections.abc import Sequence

Location = tuple[str | int, ...]


def to_path(location: Sequence[str | int]) -> str:
    """Render a document location as `operations[1].classicalTarget.bit`."""
    parts: list[str] = []

    for element in location:
        if isinstance(element, int):
            parts.append(f"[{element}]")
        elif parts:
            parts.append(f".{element}")
        else:
            parts.append(element)

    return "".join(parts)


def locate(document: object, loc: Sequence[str | int]) -> Location | None:
    """Resolve a Pydantic `loc` against a document, dropping union tags.

    Returns None when the location does not resolve, which happens for errors
    reported against a container rather than a member.
    """
    location: list[str | int] = []
    current = document

    for element in loc:
        if isinstance(current, dict):
            if element not in current:
                # A discriminated-union tag rather than a key. Skipping it is
                # correct because the tag is Pydantic's report of which branch it
                # matched, not part of the document.
                continue
            location.append(element)
            current = current[element]
        elif isinstance(current, list) and isinstance(element, int):
            if not 0 <= element < len(current):
                return None
            location.append(element)
            current = current[element]
        else:
            return None

    return tuple(location)


def pop_at(document: dict[str, object], location: Location) -> object:
    """Remove and return the value at `location`, which must name a dict key."""
    container = _descend(document, location[:-1])
    key = location[-1]

    if not isinstance(container, dict) or not isinstance(key, str):
        raise TypeError(f"{to_path(location)} does not name a field")

    return container.pop(key)


def set_at(document: dict[str, object], location: Location, value: object) -> None:
    """Put `value` back at `location`, creating missing dict parents."""
    container: object = document

    for element in location[:-1]:
        if isinstance(container, dict) and isinstance(element, str):
            container = container.setdefault(element, {})
        elif isinstance(container, list) and isinstance(element, int):
            container = container[element]
        else:
            raise TypeError(f"cannot reach {to_path(location)}")

    key = location[-1]
    if not isinstance(container, dict) or not isinstance(key, str):
        raise TypeError(f"{to_path(location)} does not name a field")

    container[key] = value


def _descend(document: dict[str, object], location: Sequence[str | int]) -> object:
    current: object = document

    for element in location:
        current = _child(current, element, location)

    return current


def _child(
    container: object, element: str | int, location: Sequence[str | int]
) -> object:
    """One step into a container, keyed or indexed according to what it is."""
    if isinstance(container, dict) and isinstance(element, str):
        return container[element]

    if isinstance(container, list) and isinstance(element, int):
        return container[element]

    raise TypeError(f"cannot reach {to_path(location)}")

File: frontend/src/serialization/paths.ts
/**
 * Locating a value inside a raw document.
 *
 * Mirrors `backend/src/phasor_workbench/serialization/paths.py`. A `Location` is
 * the sequence of keys and indices that reaches a value; `toPath` renders it in
 * the same `operations[3].targets[0]` form violations already use, so the project
 * has one vocabulary for pointing at something inside a circuit.
 *
 * **These paths are positional, and that is a constraint rather than a detail.**
 * ADR-0002 makes positions shift, so a location is only meaningful against the
 * document it was computed from. ADR-0008 section 3 is the consequence: preserved
 * fields cannot survive an edit, because editing moves the positions they name.
 */

export type Location = readonly (string | number)[];

/** Render a location as `operations[1].classicalTarget.bit`. */
export function toPath(location: Location): string {
  return location
    .map((step, index) =>
      typeof step === 'number'
        ? `[${String(step)}]`
        : index === 0
          ? step
          : `.${step}`,
    )
    .join('');
}

/** The container holding the value at `location`, or undefined if unreachable. */
function containerAt(document: unknown, location: Location): unknown {
  let current = document;

  for (const step of location.slice(0, -1)) {
    if (typeof step === 'number') {
      if (!Array.isArray(current)) return undefined;
      current = current[step];
    } else {
      if (!isRecord(current)) return undefined;
      current = current[step];
    }
  }

  return current;
}

/** Read the value at a location, or undefined if nothing is there. */
export function readAt(document: unknown, location: Location): unknown {
  const last = location.at(-1);
  const container = containerAt(document, location);
  if (last === undefined) return document;

  if (typeof last === 'number') {
    return Array.isArray(container) ? container[last] : undefined;
  }
  return isRecord(container) ? container[last] : undefined;
}

/** Remove the value at a location and return it. */
export function popAt(document: unknown, location: Location): unknown {
  const last = location.at(-1);
  const container = containerAt(document, location);
  if (last === undefined) return undefined;

  if (typeof last === 'number') {
    if (!Array.isArray(container)) return undefined;
    return container.splice(last, 1)[0];
  }
  if (!isRecord(container)) return undefined;

  const value = container[last];
  // Reflect rather than `delete container[last]`: the key is computed, and the
  // lint rule against dynamic delete exists because it deoptimizes the object.
  Reflect.deleteProperty(container, last);
  return value;
}

/**
 * Write a value at a location, creating nothing.
 *
 * A location that does not resolve is dropped rather than built out. It can only
 * arise from a document that changed shape since the location was computed, and
 * inventing the containers to reach it would graft a field onto a structure the
 * writer never saw.
 */
export function setAt(
  document: unknown,
  location: Location,
  value: unknown,
): void {
  const last = location.at(-1);
  const container = containerAt(document, location);
  if (last === undefined) return;

  if (typeof last === 'number') {
    if (Array.isArray(container) && last <= container.length) {
      container[last] = value;
    }
    return;
  }
  if (isRecord(container)) container[last] = value;
}

/** A JSON object, as opposed to an array or a primitive. */
export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * Render a value for a violation message.
 *
 * `JSON.stringify` is typed as returning `string` and returns `undefined` for
 * `undefined`, so every message that interpolates it directly can produce the
 * literal text "undefined" by accident. Naming the case is clearer than relying
 * on a `??` that the type checker believes is unreachable.
 */
export function quote(value: unknown): string {
  return value === undefined ? 'undefined' : JSON.stringify(value);
}

/**
 * Translate a JSON Pointer from a validator error into a `Location`.
 *
 * Ajv reports `/operations/0/targets/1`. Numeric segments become indices, which
 * is what lets `toPath` render them as `[1]` rather than `.1`.
 */
export function fromPointer(pointer: string): Location {
  if (pointer === '') return [];

  return pointer
    .slice(1)
    .split('/')
    .map((segment) => {
      const decoded = segment.replace(/~1/g, '/').replace(/~0/g, '~');
      return /^(0|[1-9][0-9]*)$/.test(decoded) ? Number(decoded) : decoded;
    });
}
